opentelemetry-api/src/opentelemetry/propagator/__init__.py

```diff
@@ -0,0 +1,77 @@
+import typing
+
+import opentelemetry.context.propagation.httptextformat as httptextformat
+import opentelemetry.trace as trace
+from opentelemetry.context.propagation.tracestatehttptextformat import (
+    TraceStateHTTPTextFormat,
+)
+
+
+class Propagator:
+    """Class which encapsulates propagation of values to and from context.
+
+    In contrast to using the formatters directly, a propagator object can
+    help own configuration around which formatters to use, as well as
+    help simplify the work require for integrations to use the intended
+    formatters.
+    """
+
+    def __init__(self, httptextformat_instance: httptextformat.HTTPTextFormat):
+        self._httptextformat = httptextformat_instance
+
+    def extract(
+        self, get_from_carrier: httptextformat.Getter, carrier: object
+    ) -> typing.Union[trace.SpanContext, trace.Span, None]:
+        """Load the parent SpanContext from values in the carrier.
+
+        Using the specified HTTPTextFormatter, the propagator will
+        extract a SpanContext from the carrier. If one is found,
+        it will be set as the parent context of the current span.
+
+        Args:
+            get_from_carrier: a function that can retrieve zero
+                or more values from the carrier. In the case that
+                the value does not exist, return an empty list.
+            carrier: and object which contains values that are
+                used to construct a SpanContext. This object
+                must be paired with an appropriate get_from_carrier
+                which understands how to extract a value from it.
+        """
+        span_context = self._httptextformat.extract(get_from_carrier, carrier)
+        return span_context if span_context else trace.Tracer.CURRENT_SPAN
```

Comment:
Why return the CURRENT_SPAN if extract failed? In effect, the return value can only be used as argument for `Tracer.start_span`. In that case, you might use the `trace.ParentSpan` type alias.

But I think there is an argument to be made that if there is unexpectedly no incoming span context, one would usually prefer to start a new trace instead of continuing the existing one, which would be implemented by just `return span_context`.

Reply:
sorry, you're completely right. I think I misunderstood and thought that CURRENT_SPAN was a sentinel value that was required to make sure no exceptions were raised.

I'll fix this to just be Optional[ParentSpan]

Reply:
Wouldn't the current span be null anyway? Since we pick up the trace ID from the incoming request, any span that we've created prior won't belong to the same trace.

I don't think the propagator should just return null here though. In OC we [create a new `SpanContext` with new (valid) IDs](https://github.com/c24t/opencensus-python/blob/0d004585e945292df8fe042f7415c675fff2c14c/opencensus/trace/propagation/trace_context_http_header_format.py#L42-L51) when we can't extract the span context. I don't know that we want to do the same thing here, and may want to use e.g. the invalid span instead.

Reply:
True, an invalid span context would also even better, then the return type would be simply `SpanContext`.

@toumorokoshi: The `ParentSpan` type already includes `None`, so putting it in an `Optional` is redundant.